cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -118,25 +119,38 @@ func NewFromCluster(client clientset.Interface, certificatesDir string) (*Client
 
 // Sync synchronizes client's endpoints with the known endpoints from the etcd membership.
 func (c *Client) Sync() error {
-	cli, err := clientv3.New(clientv3.Config{
-		Endpoints:   c.Endpoints,
-		DialTimeout: 20 * time.Second,
-		TLS:         c.TLS,
-	})
-	if err != nil {
-		return err
+	var cli *clientv3.Client
+	var err error
+
+	for _, endpoint := range c.Endpoints {
+		cli, err = clientv3.New(clientv3.Config{
+			Endpoints:   []string{endpoint},
+			DialTimeout: 20 * time.Second,
+			TLS:         c.TLS,
+		})
+		if err != nil {
+			return err
+		}
+
+		// check client/endpoint health before synching against it
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		_, err = cli.Get(ctx, "health")
+		cancel()
+		if err == nil || err == rpctypes.ErrPermissionDenied {
+			break
+		}
+		cli.Close()
 	}
-	defer cli.Close()
 
+	defer cli.Close()
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	err = cli.Sync(ctx)
 	cancel()
 	if err != nil {
 		return err
 	}
-	klog.V(1).Infof("etcd endpoints read from etcd: %s", strings.Join(cli.Endpoints(), ","))
-
 	c.Endpoints = cli.Endpoints()
+	klog.V(1).Infof("etcd endpoints read from etcd: %s", strings.Join(cli.Endpoints(), ","))
```

Comment:
since this log is moved after the `c.Endpoints = cli.Endpoints()` , can we change the log to do `strings.Join(c.Endpoints)` instead of doing a redundant function call?